applications/usb_uart_bridge/src/main.c

```diff
@@ -0,0 +1,260 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <device.h>
+#include <uart.h>
+#include <nrfx.h>
+#include <string.h>
+
+/* Overriding weak function to set iSerial runtime. */
+u8_t *usb_update_sn_string_descriptor(void)
+{
+	static u8_t buf[] = "PCA20035_12PLACEHLDRS";
+
+	snprintk(&buf[9], 13, "%04X%08X",
+		(uint32_t)(NRF_FICR->DEVICEADDR[1] & 0x0000FFFF)|0x0000C000,
+		(uint32_t)NRF_FICR->DEVICEADDR[0]);
+
+	return (u8_t *)&buf;
+}
+
+#define THREAD_STACKSIZE		1024
+#define UART_BUF_SIZE			58
+#define THREAD_PRIORITY			7
+
+static K_FIFO_DEFINE(usb_0_tx_fifo);
+static K_FIFO_DEFINE(usb_1_tx_fifo);
+static K_FIFO_DEFINE(uart_0_tx_fifo);
+static K_FIFO_DEFINE(uart_1_tx_fifo);
+
+struct uart_data {
+	void *fifo_reserved;
+	u8_t buffer[UART_BUF_SIZE];
+	u16_t len;
+};
+
+static struct serial_dev {
+	struct device *dev;
+	void *peer;
+	struct k_fifo *fifo;
+	struct k_sem sem;
+	struct uart_data *rx;
+} devs[4];
+
+static void uart_interrupt_handler(void *user_data)
+{
+	struct serial_dev *dev_data = user_data;
+	struct device *dev = dev_data->dev;
+	struct serial_dev *peer_dev_data = (struct serial_dev *)dev_data->peer;
+	struct uart_data **rx = &dev_data->rx;
```

Comment:
I don't think this double indirection is necessary :)

Reply:
Could you elaborate a bit on this? `dev_data->rx` needs to be both set and read several times during the function, and writing it in full every time looks a bit messy to me (see https://github.com/bjda/fw-nrfconnect-nrf/commit/81d4502d5ac14b87fb1f4364daad19897228bb00). Or did you have something else in mind?

Reply:
I am thinking, could `struct uart_data *rx = dev_data->rx` work? :)

Reply:
Ah, I tried that. The problem is that it then `rx = k_malloc(sizeof(*rx));` would not actually update `dev_data->rx`. If we were to write `dev_data->rx = k_malloc(sizeof(*rx));` to fix that, then that would not update the local `rx`. AFAICS, the local `rx` would have to be updated manually after each time `dev_data->rx` is written to (or vice versa) without this double pointer.

Reply:
I see.. another option could to just not use the extra pointer at all since it's meant to shorten the statements  but at the same time it makes them a bit tricky, perhaps just a shorter name for `dev_data` would be better:
`data_length = uart_fifo_read(dev, &(*rx)->buffer[(*rx)->len],  UART_BUF_SIZE - (*rx)->len);` vs
`data_length = uart_fifo_read(dev, dd->rx->buffer[dd->rx->len], UART_BUF_SIZE - (dd->rx->len));`

Reply:
That works for me. Implemented and called it sd (for serial device).